### PR TITLE
TIAF: Native Test Sharding Common

### PR DESCRIPTION
**Pull request by jonawals**
## What does this PR do?

This PR adds the necessary common support for native test sharding, notably:
* Adding a new label, `TIAF_sharding`, that can be added to tests registered with `ly_add_googletest` to opt-in the test target for sharding optimization (see https://github.com/aws-lumberyard-dev/o3de/pull/510 for further details of this optimization).
* Removing the sharding configuration from the TIAF config file.
* Adding serialization support for test runs and test coverage artifacts.
* Adding the necessary support to the test runner stack to facilitate another test runner type.

This PR also adds `static_assert` checks to all TIAF serializers/deserializers that make the use of key/field named constants to ensure that the number of keys always matches the number of fields.

## How was this PR tested?

Local testing and will be tested in AR.

### CODE REVIEW

**Review by hultonha — APPROVED** [review]
A lot of changes here, but as discussed offline, quite a lot of things are renames and moves. I have a couple of minor comments (I believe `_` prefix is reserved by compiler vendors so best to avoid) but otherwise looks solid to me.

File: Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp [hunk 1]
```diff
@@ -55,9 +55,11 @@ namespace TestImpact
             File,
             Discard,
             Keep,
+            // Checksum
+            _CHECKSUM_
```
[the remark this answers is not in the document]
Comment (jonawals) [reply Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp:59]: I do know that [names starting with double underscore are reserved](https://en.cppreference.com/w/cpp/language/identifiers) but a single underscore is ok (unless it falls foul of O3DE style guidelines?).

[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp:59]: From that link:

> identifiers that begin with an underscore followed by an uppercase letter

Are reserved

[the remark this answers is not in the document]
Comment (jonawals) [reply Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp:59]: Ah, I see! Ok, I'll fix and resolve this comment 👍

[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactCommandLineOptions.cpp:59]: Also stylistically a `_CHECKSUM_` doesn't match with the reset of the enum.
What is the purpose of this enum value?
Is it to track the size?
In that case the [AZ_ENUM](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Preprocessor/Enum.h#L12-L57) macro can be used.

That enables support `AZ::AzEnumTraits<CommonOptions::Fields>`, which has a `Count` member that contains the size of the enum.
It also stringifies the enum entries as well

File: Code/Tools/TestImpactFramework/Frontend/Console/Native/Code/Source/TestImpactNativeCommandLineOptions.cpp [hunk 2]
```diff
@@ -14,49 +14,38 @@
...
+    namespace NativeOptions
     {
-        enum NativeOptions
+        enum Fields
         {
             // Options
-            TestShardingPolicyKey,
             MaxConcurrencyKey,
+            // Checksum
+            _CHECKSUM_
```
Comment (amzn-ahmadkrm) [inline Code/Tools/TestImpactFramework/Frontend/Console/Native/Code/Source/TestImpactNativeCommandLineOptions.cpp:24]: If you're removing the other ```_CHECKSUM_``` here's one too!

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Artifact/Factory/TestImpactModuleCoverageFactory.cpp [hunk 3]
```diff
@@ -49,9 +49,13 @@ namespace TestImpact
                 NumberKey,
                 HitsKey,
                 SourcesKey,
-                SourceKey
+                SourceKey,
+                // Checksum
+                _CHECKSUM_
```
Comment (amzn-ahmadkrm) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Artifact/Factory/TestImpactModuleCoverageFactory.cpp:54]: as above comment

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Artifact/Factory/TestImpactTargetDescriptorFactory.cpp [hunk 4]
```diff
@@ -98,9 +98,12 @@ namespace TestImpact
             BuildDependenciesKey,
             RuntimeDependenciesKey,
             ProductionTargetTypeKey,
-            TestTargetTypeKey
+            TestTargetTypeKey,
+            // Checksum
+            _CHECKSUM_
```
Comment (amzn-ahmadkrm) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/Artifact/Factory/TestImpactTargetDescriptorFactory.cpp:103]: and here

File: Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/TestImpactChangeListSerializer.cpp [hunk 5]
```diff
@@ -16,26 +16,29 @@
 
 namespace TestImpact
 {
-    namespace ChangeListFields
+    namespace ChangeListSerializer
     {
         // Keys for pertinent JSON node and attribute names
-        constexpr const char* Keys[] =
+        constexpr const char*  Keys[] =
```
Comment (amzn-ahmadkrm) [inline Code/Tools/TestImpactFramework/Runtime/Common/Code/Source/TestImpactChangeListSerializer.cpp:22]: nit: extra space here

**Review by amzn-ahmadkrm — APPROVED** [review]
LGTM :)

**Comment by jonawals** [comment]
Thanks for all the comments! The ENUM comments have been discussed in another PR and it was decided that the changes would be in a follow up PR 👍

**Pull request merged**